test_api.py

```diff
@@ -0,0 +1,81 @@
+"""
+
+API Functionality
+=================
+
+Test the basic API functionality independent of any specific API actions.
+"""
+
+import requests
+
+from constants import API_URL
+
+
+def test_duplicate_headers(raw_options_call):
+    """ Check for duplicate headers (same key and value)
+
+    In general the same header field name may appear multiple times in a HTTP
+    response according to RFC 2616 section 4.2. But sending the exact same
+    header multiple times is always a misconfiguration. There is no need to send
+    the same value twice.
+
+    """
+
+    headers = raw_options_call.getheaders()
+    assert len(set(headers)) == len(headers), 'Duplicate response headers'
+
+
+def test_invalid_duplicate_headers(raw_options_call):
+    """ Check there are no conflicting headers
+
+    Only whitelisted header field names are allowed to appear multiple times
+    (with different values) in the response.
+
+    """
+
+    # List of header keys allowed multiple times
+    allowed_duplicate_headers = (
+        # The "Vary" conditions may be influenced by different (reverse)
+        # proxies in the delivery chain.
+        'Vary',
+    )
+
+    names = raw_options_call.headers.keys()
+    assert all(map(
+        # Check if the header is in the list of allowed headers
+        lambda x: x in allowed_duplicate_headers,
+        # in the list of duplicate headers
+        filter(lambda x: names.count(x) > 1, names)
+    ))
+
+
+def check_cors_headers(headers):
```

Comment:
I think this function, together with the tests below could be done in a single test function. As I understand it, the 200 and the 404 response has to return the same headers.

You wouldn't need the `check_cors_headers` function anymore and could just loop over two or more URLs:

```python
URLS = (
    API_URL,
    f'{API_URL}/undefined`,
)

for url in URLS:
     headers = request_headers(url)
    
     for field, value in required_cors_headers.items():
          assert headers[field] == value
```

If you want to keep the `check_cors_headers`, I would rename it to `assert_expected_cors_headers`.

Reply:
I combined it into one test.